pyworkflow/pyworkflow/node.py

```diff
@@ -23,21 +23,100 @@ def __init__(self, node_info):
 
+        self.option_replace = dict()
+        if node_info.get("option_replace"):
+            self.option_replace.update(node_info["option_replace"])
+
     def execute(self, predecessor_data, flow_vars):
-        pass
+        raise NotImplementedError()
+
+    def get_execution_options(self, flow_nodes):
+        """Replace Node options with flow variables.
+
+        If the user has specified any flow variables to replace Node options,
+        perform the replacement and return a dict with all options to use for
+        execution. If no flow variables are included, this method will return
+        a copy of all Node options unchanged.
+
+        Args:
+            flow_nodes: dict of FlowNodes used to replace options
+
+        Returns:
+            dict containing options to use for execution
+        """
+        execution_options = dict()
+
+        # TODO: Can we iterate through flow_vars instead?
+        #       If none are included, we can just return `self.options`.
+        for key, option in self.options.items():
+
+            if key in flow_nodes:
+                option.set_value(flow_nodes[key].get_replacement_value())
+
+            execution_options[key] = option
+
+        return execution_options
 
     def validate(self):
-        return True
+        """Validate Node configuration
+
+        Checks all Node options and validates all Parameter classes using
+        their validation method.
+
+        Raises:
+            ParameterValidationError: invalid Parameter value
+        """
+        for option in self.options.values():
+            option.validate()
+
+    def validate_input_data(self, num_input_data):
+        """Validate Node input data.
+
+        Checks that input data, if any, matches with required number of input
+        ports.
+
+        Args:
+            num_input_data: Number of input data passed in
+
+        Raises:
+            NodeException on mis-matched input ports/data
+        """
+        if num_input_data != self.num_in:
+            raise NodeException(
+                'execute',
+                f'{self.node_key} requires {self.num_in} inputs. {num_input_data} were provided'
+            )
+
+    def to_json(self):
+        return {
+            "name": self.name,
+            "node_id": self.node_id,
+            "node_type": self.node_type,
+            "node_key": self.node_key,
+            "data": self.data,
+            "is_global": self.is_global,
+            "option_values": self.option_values,
+            "option_replace": self.option_replace,
+        }
 
     def __str__(self):
         return "Test"
 
 
 class FlowNode(Node):
-    """FlowNode object
+    """FlowNodes object.
+
+    FlowNodes do not execute. They specify a variable name and value to pass
+    to other Nodes as a way to dynamically change other parameter values.
     """
     display_name = "Flow Control"
 
+    def execute(self, predecessor_data, flow_vars):
+        return
```

Comment:
should this raise a NotImplementedError instead?

Reply:
FlowNodes don't really have execute behavior like the other Nodes; their main logic lies in the `get_replacement_value()` method instead, so I don't think a `NotImplementedError` is needed here.

This also allows for the children (only the `StringNode`) to avoid adding empty `execute()` methods.